eclipse/src/saros/ui/wizards/AddProjectToSessionWizard.java

```diff
@@ -666,4 +680,79 @@ private void displaySaveDirtyEditorsDialog(final List<IEditorPart> dirtyEditors)
       for (IEditorPart editor : dirtyEditors) editor.doSave(new NullProgressMonitor());
     }
   }
+
+  private void storeCurrentProjectNameMapping(final JID jid) {
+    Map<String, String> currentProjectNameMapping = new HashMap<>();
+
+    for (final ProjectNegotiationData data : negotiation.getProjectNegotiationData()) {
+      final String projectID = data.getProjectID();
+      currentProjectNameMapping.put(
+          data.getProjectName(), namePage.getTargetProjectName(projectID));
+    }
+
+    setNewProjectNameMapping(jid, currentProjectNameMapping);
+  }
+
+  private Map<String, String> getLastProjectNameMapping(final JID jid) {
+    Map<String, String> result = new HashMap<>();
+
+    final String mapping = preferenceStore.getString(projectMappingStorePrefix + jid.getBase());
+
+    if (mapping == null || mapping.isEmpty()) return result;
+
+    String remoteName = null;
+    String localName = null;
+
+    for (int i = 0, j = 0; j < mapping.length(); j++) {
+
+      boolean isLastCharacter = j == mapping.length() - 1;
+
+      if (mapping.charAt(j) == ':' || isLastCharacter) {
+        if (j == 0) // corrupted data ...
+        return result;
+
+        if (!isLastCharacter && mapping.charAt(j - 1) == '\\') continue;
+
+        if (isLastCharacter) j++;
+
+        final String name = mapping.substring(i, j);
+
+        i = j + 1;
+
+        if (remoteName == null) remoteName = name;
+        else localName = name;
+
+        if (remoteName != null && localName != null) {
+          result.put(remoteName.replace("\\:", ":"), localName.replace("\\:", ":"));
+          remoteName = localName = null;
+        }
+      }
+    }
+
+    return result;
+  }
+
+  private void setNewProjectNameMapping(
+      final JID jid, final Map<String, String> projectNameMapping) {
+
+    final Map<String, String> lastProjectNameMapping = getLastProjectNameMapping(jid);
+
+    for (final Entry<String, String> entry : projectNameMapping.entrySet())
+      lastProjectNameMapping.put(entry.getKey(), entry.getValue());
```

Comment:
Just adding the new mapping to the existing mapping could lead to some weird states.
If I were to
- accept a project now and set up the mapping A->A, B->B, and C->C
- afterwards, accept a new session (with the same host) and set up the mapping B->C
- afterwards, accept a new session with the initial three projects

This would lead to the automatic setup suggesting to map two resources to C (A->A, B->C, C->C).

In general, as this call potentially overwrites existing mappings, it could lead to mixing the mappings of different sessions. A solution would be to hold mappings for project combinations and always trying to apply the largest held mappings (you could still also try to combine non-overlapping mappings). But this would make the logic more complicated and would increase the saved data size.

As already stated, I find it hard to construct a behavior with this logic that covers the most important use-cases without leading to too much confusion for the users in the other (corner) cases.

Reply:
I see three options here.
1. Store set of sets (as you already mentioned an overkill)
2. Refuse to update the specific entry if it already exists
3. Replace the mapping and remove affected stuff, i.e if I map B->C then I would remove the C->C mapping.

If would prefer option 3.

Reply:
Option three would probably be the best compromise. 

But it could still lead to unexpected situations. When you replace the second step in the example with mapping B->D, the mapping in step three would be legal but still not what the user expects (B->D instead of B->B).
Either way, there won't be a perfect solution, so we have to somehow get the user to check the tabs again. But this also somewhat reduces the usefulness of the feature if the user still has to check everything (at least when there were other sessions in-between using different mappings).

How about holding whether a mapping was overwritten and only then notifying the user/marking these tabs? This could minimize the needed interaction by the user while reducing the probability of unexpected outcomes.